overlord/snapstate/snapstate.go

```diff
@@ -579,6 +579,13 @@ func validateFeatureFlags(st *state.State, info *snap.Info) error {
 		if !flag {
 			return fmt.Errorf("experimental feature disabled - test it by setting 'experimental.user-daemons' to true")
 		}
+		// Ubuntu 14.04's systemctl does not support the
+		// arguments needed to enable user session
+		// units. Further more, it does not ship with a
+		// systemd user instance.
+		if release.ReleaseInfo.ID == "ubuntu" && release.ReleaseInfo.VersionID == "14.04" {
```

Comment:
Thanks for starting this.

I think this should be somewhat different. Either systemd or release package should know if this is supported and we should call it here. From the top of my head I can say this would also need to include CentOS 7 and Amazon Linux 2 that entirely disable, at compile time, systemd --user

Reply:
As discussed on IRC, one difference here is that the `systemctl` in CentOS 7/Amazon Linux 2 will let the snaps be successfully installed: they've just disabled the startup of a the user systemd instance.

Also, a server operating system is unlikely to start a user session either way.  Would we want to block install on Ubuntu Core systems?

Reply:
@jhenstridge I see your point but my point is that this piece of code should ask an oracle instead of blacklisting one specific system.

As for core, I think on core20 it should not be disabled.

Reply:
Isn't `core20` in the same boat as previous Ubuntu Core releases?  It supports user daemons, but in practice there is never a user session.

The impetus for this was that installing such a snap on 14.04 will fail later on in the install process. That seems distinct from cases where the install will succeed, but the snap may not be useful.  If we do that, we're in the realm of blocking install of desktop snaps on headless Ubuntu Server systems.

Reply:
I think I agree with @zyga point that the predicate itself should be a function on the release package.

Reply:
For what it is worth, this is the relevant patch CentOS 7 applies to systemd:

https://git.centos.org/rpms/systemd/blob/c7/f/SOURCES/0004-remove-user-.service.patch

It removes prevents `/lib/systemd/system/user@.service` from being installed, and patches logind so it doesn't try to start that service.  All the code supporting a user instance within systemd itself and the systemctl utility is still present.

That's why I said it is closer to a system with no running user sessions than systems with an ancient systemd like Ubuntu 14.04.